File: cpp/test/TestCrashers.cpp

#include "Tokenizer.h"
#include "ParserSession.h"
#include "TokenEnumRegistration.h"

#include "gtest/gtest.h"

class ParserSession;

using ParserSessionPtr = ParserSession *;


class CrashTest : public ::testing::Test {
protected:
    
    static ParserSessionPtr session;
    
    static void SetUpTestSuite() {
        
        session = new ParserSession();
    }
    
    static void TearDownTestSuite() {
        
        delete session;
    }
    
    void SetUp() override {
        
    }
    
    void TearDown() override {
        
    }
};

ParserSessionPtr CrashTest::session;


TEST_F(CrashTest, Crash0_tokens) {
    
    const unsigned char arr[] = {'1', '\\', '\n'};
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(arr, 3, nullptr, opts);

    auto policy = TOPLEVEL;

    auto Tok = Tokenizer_currentToken(session, policy);

    EXPECT_EQ(Tok, Token(TOKEN_INTEGER, session->start + 0, 1, Source(SourceLocation(1, 1), SourceLocation(1, 2))));
    
    Tokenizer_nextToken(session, policy);
    
    Tok = Tokenizer_currentToken(session, policy);

    EXPECT_EQ(Tok, Token(TOKEN_ENDOFFILE, session->start + 1, 3, Source(SourceLocation(1, 2), SourceLocation(2, 1))));

    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
    
    session->deinit();

    SUCCEED();
}

TEST_F(CrashTest, Crash1) {
    
    const unsigned char arr[] = {'1', ':', ':', '*', '\\', '\r', '\n'};
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(arr, 7, nullptr, opts);
    
    auto N = session->concreteParse();
    
    session->releaseNode(N);
    
    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
    
    session->deinit();
    
    SUCCEED();
}

#if 0
TEST_F(CrashTest, StackOverflow1) {
    
    unsigned char arr[1600];
    for (auto i = 0; i < 1600 ; i++){
        arr[i] = '(';
    }
    
    TheParserSession->init(arr, 1600, nullptr, SOURCECONVENTION_LINECOLUMN);
    
    auto N = TheParserSession->parseExpressions();
    
    TheParserSession->releaseNode(N);
    
    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
    
    TheParserSession->deinit();
    
    SUCCEED();
}
#endif // #if 0

TEST_F(CrashTest, Crash2) {
    
    const unsigned char arr[] = {'\\', ':', 'f', 'e', 'f', 'f'};
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(arr, 6, nullptr, opts);

    auto N = session->concreteParse();

    session->releaseNode(N);

    EXPECT_EQ(session->nonFatalIssues.size(), 1u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
    
    session->deinit();

    SUCCEED();
}

TEST_F(CrashTest, Crash3) {
    
    const unsigned char arr[] = {'a', ':', 'b', '~', '1', ':', '2'};
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(arr, 7, nullptr, opts);

    auto N = session->concreteParse();

    session->releaseNode(N);

    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
    
    session->deinit();

    SUCCEED();
}

TEST_F(CrashTest, Crash4) {
    
    const unsigned char arr[] = {'\\', '[', 'I', 'n', 't', 'e', 'g', 'r', 'a', 'l', ']', '\\', '[', 'S', 'u', 'm', ']'};
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(arr, 17, nullptr, opts);

    auto N = session->concreteParse();

    session->releaseNode(N);

    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
    
    session->deinit();

    SUCCEED();
}

TEST_F(CrashTest, Crash5) {
    
    const unsigned char arr[] = {'{', '\t', '1', '\\', '\n', '^'};
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_SOURCECHARACTERINDEX;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(arr, 6, nullptr, opts);

    auto N = session->concreteParse();

    session->releaseNode(N);

    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
    
    session->deinit();

    SUCCEED();
}

File: crate/src/tests/test_crashers.rs
use crate::{
    source::{ByteSpan, TOPLEVEL},
    src,
    token::Token,
    token_enum::TokenEnum::*,
    tokenizer::{Tokenizer_currentToken, Tokenizer_nextToken},
    EncodingMode, FirstLineBehavior, ParserSession, SourceConvention, DEFAULT_TAB_WIDTH,
};


#[test]
fn CrashTest_Crash0_tokens() {
    let bufAndLen: &[u8] = &[b'1', b'\\', b'\n'];

    let mut session = ParserSession::new(
        bufAndLen,
        SourceConvention::LineColumn,
        DEFAULT_TAB_WIDTH,
        FirstLineBehavior::NotScript,
        EncodingMode::Normal,
    );

    let policy = TOPLEVEL;

    let mut Tok = Tokenizer_currentToken(&mut session.tokenizer, policy);

    assert_eq!(
        Tok,
        Token::new3(TOKEN_INTEGER, ByteSpan::new(0, 1), src!(1:1-1:2))
    );

    Tokenizer_nextToken(&mut session.tokenizer, policy);

    Tok = Tokenizer_currentToken(&mut session.tokenizer, policy);

    assert_eq!(
        Tok,
        Token::new3(TOKEN_ENDOFFILE, ByteSpan::new(1, 2), src!(1:2-2:1))
    );

    assert_eq!(session.nonFatalIssues().len(), 0);
    assert_eq!(session.fatalIssues().len(), 0);
}

#[test]
fn CrashTest_Crash1() {
    let bufAndLen = &[b'1', b':', b':', b'*', b'\\', b'\r', b'\n'];

    let mut session = ParserSession::new(
        bufAndLen,
        SourceConvention::LineColumn,
        DEFAULT_TAB_WIDTH,
        FirstLineBehavior::NotScript,
        EncodingMode::Normal,
    );

    let _ = session.concrete_parse_expressions();

    assert_eq!(session.nonFatalIssues().len(), 0);
    assert_eq!(session.fatalIssues().len(), 0);
}

/*
#if 0
#[test]
fn CrashTest_StackOverflow1() {

    unsigned char arr[1600];
    for (let i = 0; i < 1600 ; i++){
        arr[i] = '(';
    }

    let bufAndLen = BufferAndLength(arr, 1600);

    TheParserSession->init(bufAndLen, nullptr, SourceConvention::LineColumn);

    let N = TheParserSession->parseExpressions();

    TheParserSession->releaseNode(N);

    assert_eq!(session.nonFatalIssues().len(), 0);
    assert_eq!(session.fatalIssues().len(), 0);

    TheParserSession->deinit();
}
#endif // #if 0
*/

#[test]
fn CrashTest_Crash2() {
    let arr: &[u8] = &[b'\\', b':', b'f', b'e', b'f', b'f'];

    let bufAndLen = arr;

    let mut session = ParserSession::new(
        bufAndLen,
        SourceConvention::LineColumn,
        DEFAULT_TAB_WIDTH,
        FirstLineBehavior::NotScript,
        EncodingMode::Normal,
    );

    let _ = session.concrete_parse_expressions();

    assert_eq!(session.nonFatalIssues().len(), 1);
    assert_eq!(session.fatalIssues().len(), 0);
}

#[test]
fn CrashTest_Crash3() {
    let arr: &[u8] = &[b'a', b':', b'b', b'~', b'1', b':', b'2'];

    let bufAndLen = arr;

    let mut session = ParserSession::new(
        bufAndLen,
        SourceConvention::LineColumn,
        DEFAULT_TAB_WIDTH,
        FirstLineBehavior::NotScript,
        EncodingMode::Normal,
    );

    let _ = session.concrete_parse_expressions();

    assert_eq!(session.nonFatalIssues().len(), 0);
    assert_eq!(session.fatalIssues().len(), 0);
}

#[test]
fn CrashTest_Crash4() {
    let arr: &[u8] = &[
        b'\\', b'[', b'I', b'n', b't', b'e', b'g', b'r', b'a', b'l', b']', b'\\', b'[', b'S', b'u',
        b'm', b']',
    ];

    let bufAndLen = arr;

    let mut session = ParserSession::new(
        bufAndLen,
        SourceConvention::LineColumn,
        DEFAULT_TAB_WIDTH,
        FirstLineBehavior::NotScript,
        EncodingMode::Normal,
    );

    let _ = session.concrete_parse_expressions();

    assert_eq!(session.nonFatalIssues().len(), 0);
    assert_eq!(session.fatalIssues().len(), 0);
}

#[test]
fn CrashTest_Crash5() {
    let arr: &[u8] = &[b'{', b'\t', b'1', b'\\', b'\n', b'^'];

    let bufAndLen = arr;

    let mut session = ParserSession::new(
        bufAndLen,
        SourceConvention::CharacterIndex,
        DEFAULT_TAB_WIDTH,
        FirstLineBehavior::NotScript,
        EncodingMode::Normal,
    );

    let _ = session.concrete_parse_expressions();

    assert_eq!(session.nonFatalIssues().len(), 0);
    assert_eq!(session.fatalIssues().len(), 0);
}
